### ISSUE TITLE
[C++ API] use of `std::function` ABI may not be stable 

### ISSUE THREAD

**Issue by jon-chuang** [ray-project/ray#21243]
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

These are the function signatures that acts as the FFI between the application code and ray runtime (C++ worker runtime)
```cpp
using RemoteFunction = std::function<msgpack::sbuffer(const ArgsBufferList &)>;
using RemoteMemberFunction =
    std::function<msgpack::sbuffer(msgpack::sbuffer *, const ArgsBufferList &)>;
```

However, it is known that `std::function` does not have a stable ABI, e.g. https://stackoverflow.com/questions/60291845/undocumented-abi-changes-of-stdfunction-between-gcc-4-and-gcc-5-6-7-8-9-how-t
The same could be true of `msgpack::sbuffer(const ArgsBufferList &)`...

### Use case

Not breaking unexpectedly

### Recommendation

Convert function type signatures to be C-ABI-compatible, which is more stable and the typical standard for FFI.

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

**Comment by jon-chuang** [comment]
Furthermore, the C++ worker ABI also exposes C++ types like HashMap. For some ideas on how to convert the hashmap, they could be to cast function names as byte arrays, and use those C-ABI compatible byte-arrays as the key for the fast function lookup, thus exposing only C-ABI compatible function signatures.

@SongGuyang @kfstorm @jovany-wang

**Comment by SongGuyang** [comment]
@jon-chuang Thank you for add this issue! Looks good to me. I agree that we should make C++ API headers more compatible and simple. Maybe we should do a review for all the headers. @qicosmos please pay attention to this issue.

**Comment by qicosmos** [comment]
Actually it's not a problem for ray, ray use c++17 now(at least gcc7.2 above), that means if you want to use ray you must use gcc7+ compiler, the ABI issue will occur when use library built with lower gcc version(below), because the std::string and std::function implement changed on higher gcc, gcc use _GLIBCXX_USE_CXX11_ABI flag to solve the issue.
Because ray always run on gcc7 above, so there is no ABI issue.

**Comment by jon-chuang** [comment]
How about hashmap ABI?

**Comment by qicosmos** [comment]
> How about hashmap ABI/implementation?

I think it's similar with string and function, we use gcc7 above compiler, no abi problem.

**Comment by jon-chuang** [comment]
Hmm it's not my decision to make, since it's ultimately about priority and I agree it won't result in breakage in most cases, while one could also probably fix breakage by configuring the right build and absl release version for the application code, and I admit I don't understand the story for C++ compiler ABI guarantees, but maybe choosing this path is an unnecessary risk that could complicate future development, whereas it could be solved with a small amount of refactoring?

Furthermore, what if someone has a codebase building with clang? Can you still guarantee ABI compat then?

**Comment by qicosmos** [comment]
C++ can't solve abi issue between different compiler even different version of the same compiler now.
The best way is : compile ray with the same compile to keep ABI compatible, no other way now.

**Comment by jon-chuang** [comment]
Hmm I don't think I believe this. For instance, rust API is compiled by a completely different compiler but there aren't issues expected, the reason is because the application code interfaces with the runtime exclusively via a wrapped task function signature, which can be guaranteed to interoperate as long as that ABI is preserved (i.e. by declaring as `extern "C"` to generate C ABI), and as long as the serialization format for arguments and return values is consistent for different types. 

This is just what I understand to be standard FFI practices, even if it is FFI from C++ to C++.

**Comment by qicosmos** [comment]
Yes, use extern c is a solution, it's a c solution, that also means you abandon c++,  all the user remote function must be exported with extern c, can't use c++ string, function, c++ classes, stl container any more, that's bad for c++ users.

If we need to provide c api for c api users, it's another story.

**Comment by jon-chuang** [comment]
I don't think you understand what an FFI is.

**Comment by jon-chuang** [comment]
Extern C is not a C solution. It is a generic FFI solution that is industry standard practice

**Comment by qicosmos** [comment]
Extern c means you can't use c++ classed, containers, i think c api is another story.

**Comment by jon-chuang** [comment]
Please read the C++ API implementation properly. 

All user remote functions already follow one of two fixed type signatures listed above. Both of which have no usage of STL.

Please try to understand how FFI is done for complex type signatures before commenting further.

**Comment by qicosmos** [comment]
> Please read the C++ API implementation properly.
> 
> All user remote functions already follow one of two fixed type signatures listed above. Both of which have no usage of STL.
> 
> Please try to understand how FFI is done for complex type signatures before commenting further.

The user function is free for users, they can pass c++ stl container and string, if use extern c, how to pass the c++ arguments? use "const char* , size_t" instead of string, use what instead of std::vector arguments? void*?
Don't you think it's very hard to use?

**Comment by jon-chuang** [comment]
The user defined remote function's signature is not the FFI. I have already mentioned above what the FFI currently is in the original post. 

hint: RemoteFunction and RemoteMemberFunction.

**Comment by jon-chuang** [comment]
This is how to pass.

```
Extern "C" remote_function(args: FFISafeBuffer) {
   auto arg = msgpack::Deserialize<std::vector<int>>(args[0]);
   myfavouritearbitrartcppfunction(arg);
}
```

**Comment by qicosmos** [comment]
Yes, i mean it's not convenient for c++ users, because the user have to change the original c++ arguments to c api acceptable arguments. The current API no such restrictions, so i said it's for c++ users, if we need c APIs, we can provide new c APIs.

**Comment by jon-chuang** [comment]
Yes, but I'm saying we can wrap the user function for them. For instance:

```cpp
extern "C" find_and_apply_remote_function(msgpack::sbuffer key, msgpack::sbuffer args) msgpack::sbuffer {
  return function_manager.get(key)(args);
}
```

**Comment by qicosmos** [comment]
> 

Actually we have already use extern c for the exported functions by boost.dll BOOST_DLL_ALIAS in task_executor.h.

**Comment by jon-chuang** [comment]
I don't think that makes it extern "C"? I think it just does not mangle the names?

**Comment by jon-chuang** [comment]
Some additional evidence that one might not be able to rely on C++11 ABI
![Screenshot_20211224-145132_Chrome](https://user-images.githubusercontent.com/9093549/147326578-5e97db50-fcc9-4547-991c-80d542c9d0c9.jpg)

**Comment by SongGuyang** [comment]
@mwtian What do you think about this discuss?

**Comment by mwtian** [comment]
To my knowledge, user code -> Ray C++ API does not offer an API stability guarantee yet, much less an ABI guarantee. I don't think we need to worry about ABI guarantee yet.

A clear C++ ABI guarantee is not trivial to maintain. Build system and [C++ language standard](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p1018r6.html#abi) have to be taken into account. It would be hard for users to follow as well. C FFI for C++ worker can maintain ABI stability, but it imposes restrictions on the C++ API, e.g. [no tempated args like std::vector<>](https://stackoverflow.com/questions/52367382/is-it-possible-to-use-template-arguments-in-extern-c-functions). I think requiring users to rebuild their application with each new version of Ray C++ worker is the right choice for now.

**Comment by jon-chuang** [comment]
Hmm, I agree its not an urgent priority. However:

> but it imposes restrictions on the C++ API, e.g. no tempated args like std::vector<>.

Yes, but its easy to fix with serialization, which in this case is quite lightweight.

>  but it seems easier to document and less likely to go out of date compared to an ABI stability guarantee.

I agree the FFI might change, but I think C ABI is a good starting point to build future development upon. 

I think the FFI breakage with C ABI is very explicit - as it requires actual code changes, not just build system change, as could happen with C++ ABI. 

Hence, I don't think it is harder to document or will go out of date, as the `extern "C"` FFI signature is self-documenting.

Both changing the `extern "C"` FFI or C++ FFI would also warrant warning users for the breaking-changes in the release. 

Perhaps since the C++ FFI is also currently unstable, it might be the best time to make some changes, so that one doesn't have to change the FFI later on when users expect more stability.

---

The nice thing about  `extern "C"` FFI is that one does not ever need to worry about recompiling application `.so` files even if user upgrades ray installation (barring FFI changes). 

---

Hence, I want to see if we can get consensus on whether in the long-term, we want to move to a C-ABI.

I belive I could contribute to this process in the future, and I don't think it's very difficult to achieve.

**Comment by mwtian** [comment]
@jon-chuang If this is about a Ray-internal C API which wraps core worker C++ API, that other language workers (Rust, even C++) can build upon, it would be more feasible. The benefit of the C API (Core worker C++ API -> C API -> language workers) is to provide compatibility with multiple languages' FFI facilities, instead of ABI stability (there is no intention to offer core worker **API** stability yet). @iycheng has investigated this before too.

The value of C API on top of C++ API seems less obvious. We can be more productive by having a minimum example (start Ray, run a task, get the result etc) of how the proposed set of C APIs would look like, and how users should interact with them. Then issues about API ergonomics, efficiency and ABI stability would be clearer.

> Hence, I want to see if we can get consensus on whether in the long-term, we want to move to a C-ABI.

Where are we moving from, the existing C++ API?
